blocksuite/affine/shared/src/viewport-renderer/renderer-utils.ts

```diff
@@ -23,58 +27,104 @@ export function syncCanvasSize(canvas: HTMLCanvasElement, host: HTMLElement) {
   canvas.style.pointerEvents = 'none';
 }
 
+function clientToModelCoord(
+  viewport: ViewportRecord,
+  clientCoord: [number, number]
+) {
+  const { left, top, viewportX, viewportY, zoom, viewScale } = viewport;
+
+  const [clientX, clientY] = clientCoord;
+  const viewportInternalX = clientX - left;
+  const viewportInternalY = clientY - top;
+  const modelX = viewportX + viewportInternalX / zoom / viewScale;
+  const modelY = viewportY + viewportInternalY / zoom / viewScale;
+  return [modelX, modelY];
+}
+
+function getParagraphs(host: EditorHost) {
+  const gfx = host.std.get(GfxControllerIdentifier);
+  const doc = host.doc;
+  const surface = gfx.surface;
+
+  const surfaceChildren = surface?.children || [];
+  const rootChildren = doc.root?.children || [];
+  const nonSurfaceChildren = rootChildren.filter(c => c.id !== surface?.id);
+  const blocks = [...surfaceChildren, ...nonSurfaceChildren];
+  const components = blocks
+    .map(block => gfx.view.get(block.id))
+    .filter(Boolean) as GfxBlockComponent[];
```

Comment:
You can use `gfx.gfxElements.filter(e => e instanceof GfxBlockElementModel)` here

Reply:
Thanks for the hint!